### PR TITLE
Refactor and bug fixes

### PR DESCRIPTION
**Pull request by belanglos**
This PR contains a lot of refactoring and bug fixes

* KUI-1176
* fix: remove fade in/out
* refactor context, make it read-only and object instead of array containing an object
* refactor call to ug-rest-api
* return 400 if given courseCode has wrong length

### CODE REVIEW

File: server/controllers/index.js [hunk 1]
```diff
@@ -11,4 +11,6 @@ module.exports = {
   noCourse: require('./noCourseCtrl'),
   StatisticsCtrl: require('./statisticsCtrl'),
   TimeTableApi: require('./timeTableApiCtrl'),
+  Kopps: require('./koppsCourseCtrl'),
+  Employees: require('./employeesCtrl'),
```
Comment (karlandindrakryggen) [inline server/controllers/index.js:15]: Great that moved the API endpoints out from courseCtrl to to own controllers!

File: public/js/app/hooks/useSemesterRoundState.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    rounds[semester].length === 1
+
+  const activeSemesterOnlyHasOneRound = useMemo(
+    () =>
+      roundList !== undefined &&
+      Object.hasOwnProperty.call(roundList, selectedSemester) &&
+      roundList[selectedSemester] &&
+      roundList[selectedSemester].length === 1,
+    [roundList, selectedSemester]
+  )
```
[the remark this answers is not in the document]
Comment (karlandindrakryggen) [reply public/js/app/hooks/useSemesterRoundState.js:40]: Can `semesterOnlyHasOneRound` be named something else to make it clear? Not sure how to name it though.

**Review by karlandindrakryggen — APPROVED** [review]
Great work! 

I have some small questions/suggestions, but nothing important, so I think it's ok to merge ✅

File: server/apiCalls/filteredData.js
[the remark this answers is not in the document]
Comment (belanglos) [reply server/apiCalls/filteredData.js:1]: Makes sense!

File: public/css/kursinfo-web.scss [hunk 3]
```diff
@@ -140,21 +140,14 @@ sup {
...
-    0% {
-      opacity: 1;
-    }
-    50% {
-      opacity: 0;
-    }
-    100% {
-      opacity: 1;
-    }
+  .right-column-header {
```
[the remark this answers is not in the document]
Comment (belanglos) [reply public/css/kursinfo-web.scss:143]: True! Good catch. I think I must have made a mistake while rebasing removing the fade in/out on top of style.

File: public/js/app/hooks/useSemesterRoundState.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (belanglos) [reply public/js/app/hooks/useSemesterRoundState.js:40]: Uff! Very good catch, thank you! ⭐

File: server/ssr-context/createServerSideContext.js [hunk 4]
```diff
@@ -8,41 +8,8 @@ function setBrowserConfig(config, paths, hostUrl) {
...
-    semesterSelectedIndex: 0,
-    roundSelectedIndex: 0,
-    courseData: {
-      courseInfo: {
-        course_application_info: [],
-      },
-      syllabusSemesterList: [],
-    },
+    isCancelledOrDeactivated: false, // Done
+    initiallySelectedRoundIndex: undefined, // Source // Done
```
[the remark this answers is not in the document]
Comment (belanglos) [reply server/ssr-context/createServerSideContext.js]: [faab00f](https://github.com/KTH/kursinfo-web/pull/339/commits/faab00f893938320b06bb97732351592851ed53a)

File: server/apiCalls/filteredData.js
[the remark this answers is not in the document]
Comment (belanglos) [reply server/apiCalls/filteredData.js:1]: [222b7e6](https://github.com/KTH/kursinfo-web/pull/339/commits/222b7e617332eb3622692d7494c204932136f999)

File: public/css/kursinfo-web.scss (hunk 3 above)
[the remark this answers is not in the document]
Comment (belanglos) [reply public/css/kursinfo-web.scss]: [2634e62](https://github.com/KTH/kursinfo-web/pull/339/commits/2634e623aa0085023f9cc7b3bcf73ccfd7228c23)

File: public/js/app/hooks/useSemesterRoundState.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (belanglos) [reply public/js/app/hooks/useSemesterRoundState.js:36]: [9af196e](https://github.com/KTH/kursinfo-web/pull/339/commits/9af196e95709eb1600888f2667869ed0f7765d12)

**Comment by belanglos** [comment]
Addressed the issues and also renamed roundList (which confusingly enough is an object) to roundsBySemester.

**Pull request merged**